src/Core/Form/IdentifiableObject/DataProvider/ProductFormDataProvider.php

```diff
@@ -156,6 +175,42 @@ private function addShortcutData(array $productData): array
         return $productData;
     }
 
+    /**
+     * @param ProductForEditing $productForEditing
+     *
+     * @return array<string, int|string>
+     */
+    private function extractVirtualProductFileData(ProductForEditing $productForEditing): array
```

Comment:
Could you add unit test for this part, please ?

Reply:
sorry i've tried, but to be honest I don't understand what the hell is going on in that "unit" test anymore. It is a perfect example of how to not use unit tests. The test is 10x more complicated than the actual class. Need to do something about that :smile:

Reply:
okay so I fixed the failing test, but didn't add virtual product file as it requires filesystem mocking and Im lost in this big "unit" test. So we need to split ProductFormDataProvider into multiple simplier providers and test them one by one. Added todo in test.